Add DPDK version 24.11.3

edbird opened the pull request:
- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [X] SHA512s are updated for each updated download.
- [X] The "supports" clause reflects platforms that may be fixed by this new version.
- [X] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [X] Any patches that are no longer applied are deleted from the port's directory.
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [X] Only one version is added to each modified port's versions file.

Commit:
Add DPDK version 24.11.3

edbird commented:
> You could add the port directory as an overlay-port in your build and see if that succeeds

Here's what I have done:

Created the directory `vcpkg-overlay/ports/dpdk` which is relative to my project root. (Directory containing vcpkg.json)

Copied `~/vcpkg/ports/dpdk/portfile.cmake` to `vcpkg-overlay/ports/dpdk`.

Edit this file, change `vcpkg_from_github` block to this:

```
vcpkg_from_github(
  OUT_SOURCE_PATH SOURCE_PATH
  REPO DPDK/dpdk
  REF "v24.11.3"
  SHA512 0
  HEAD_REF main
)
```

Note I removed the list of patches here. Not sure if I should have done this.

When I try to build, I see the same error as I was intending to fix. That is

```
../src/v24.07-c7d2488347.clean/drivers/net/gve/base/gve_osdep.h:41:20: error: conflicting types for ‘__be64’; have ‘rte_be64_t’ {aka ‘long unsigned int’}                                                                                          41 | typedef rte_be64_t __be64;
```

Is what I have attempted to do here the correct thing to do to test the change?

Commit:
Update SHA512 and version string

Commit:
Update SHA512

Commit:
vcpkg x-add-version dpdk --overwrite-version

edbird commented on versions/d-/dpdk.json:
```diff
@@ -1,5 +1,10 @@
+      "git-tree": "b287e404643aeab5bb5db084c142789df187844e",
+      "version-string": "24.11.3",
+      "port-version": 0
+    },
```
I think I have now completed this step. I note that the `versions/baseline.json` file has updated.

edbird commented:
This is the output from my test project when I run `cmake --preset=vcpkg`

```
-- Running vcpkg install
Fetching registry information from https://github.com/microsoft/vcpkg (HEAD)...
Detecting compiler hash for triplet x64-linux...
Compiler found: /usr/bin/c++
The following packages will be built and installed:
  * bzip2[core,tool]:x64-linux@1.0.8#6 -- git+https://github.com/microsoft/vcpkg@2d029da682847c5ebdc54e4dbea001331a02207e
    dpdk:x64-linux@24.07 -- git+https://github.com/microsoft/vcpkg@a4d23173aedf2ba7b16305a9bb0b110bd8b05026
```

It still picks up the wrong version of dpdk. (24.07). Why is this? I tried deleting the downloaded file `DPDK-dpdk-v24.07.tar.gz` and cleared the build directory.

edbird commented:
I tried to use classic mode and seem to have managed to get further.

```
vcpkg install dpdk
```

It now picks up the correct version.

```
The following packages will be built and installed:
    dpdk:x64-linux@24.11.3
```

I removed all patches.

I am now encountering this error:

```
[98/2667] /usr/bin/cc -fPIC -g  -o lib/librte_eal.so.25.0 lib/librte_eal.a.p/eal_common_eal_common_bus.c.o lib/librte_eal.a.p/eal_common_eal_common_class.c.o lib/librte_eal.a.p/eal_common_eal_common_config.c.o lib/librte_eal.a.p/eal_common_eal_common_debug.c.o lib/librte_eal.a.p/eal_common_eal_common_dev.c.o lib/librte_eal.a.p/eal_common_eal_common_devargs.c.o lib/librte_eal.a.p/eal_common_eal_common_errno.c.o lib/librte_eal.a.p/eal_common_eal_common_fbarray.c.o lib/librte_eal.a.p/eal_common_eal_common_hexdump.c.o lib/librte_eal.a.p/eal_common_eal_common_interrupts.c.o lib/librte_eal.a.p/eal_common_eal_common_launch.c.o lib/librte_eal.a.p/eal_common_eal_common_lcore.c.o lib/librte_eal.a.p/eal_common_eal_common_lcore_var.c.o lib/librte_eal.a.p/eal_common_eal_common_mcfg.c.o lib/librte_eal.a.p/eal_common_eal_common_memalloc.c.o lib/librte_eal.a.p/eal_common_eal_common_memory.c.o lib/librte_eal.a.p/eal_common_eal_common_memzone.c.o lib/librte_eal.a.p/eal_common_eal_common_options.c.o lib/librte_eal.a.p/eal_common_eal_common_string_fns.c.o lib/librte_eal.a.p/eal_common_eal_common_tailqs.c.o lib/librte_eal.a.p/eal_common_eal_common_thread.c.o lib/librte_eal.a.p/eal_common_eal_common_timer.c.o lib/librte_eal.a.p/eal_common_eal_common_trace_points.c.o lib/librte_eal.a.p/eal_common_eal_common_uuid.c.o lib/librte_eal.a.p/eal_common_malloc_elem.c.o lib/librte_eal.a.p/eal_common_malloc_heap.c.o lib/librte_eal.a.p/eal_common_rte_bitset.c.o lib/librte_eal.a.p/eal_common_rte_malloc.c.o lib/librte_eal.a.p/eal_common_rte_random.c.o lib/librte_eal.a.p/eal_common_rte_reciprocal.c.o lib/librte_eal.a.p/eal_common_rte_service.c.o lib/librte_eal.a.p/eal_common_rte_version.c.o lib/librte_eal.a.p/eal_common_eal_common_dynmem.c.o lib/librte_eal.a.p/eal_common_eal_common_cpuflags.c.o lib/librte_eal.a.p/eal_common_eal_common_hypervisor.c.o lib/librte_eal.a.p/eal_common_eal_common_proc.c.o lib/librte_eal.a.p/eal_common_eal_common_trace.c.o lib/librte_eal.a.p/eal_common_eal_common_trace_ctf.c.o lib/librte_eal.a.p/eal_common_eal_common_trace_utils.c.o lib/librte_eal.a.p/eal_common_hotplug_mp.c.o lib/librte_eal.a.p/eal_common_malloc_mp.c.o lib/librte_eal.a.p/eal_common_rte_keepalive.c.o lib/librte_eal.a.p/eal_unix_eal_debug.c.o lib/librte_eal.a.p/eal_unix_eal_file.c.o lib/librte_eal.a.p/eal_unix_eal_filesystem.c.o lib/librte_eal.a.p/eal_unix_eal_firmware.c.o lib/librte_eal.a.p/eal_unix_eal_unix_memory.c.o lib/librte_eal.a.p/eal_unix_eal_unix_thread.c.o lib/librte_eal.a.p/eal_unix_eal_unix_timer.c.o lib/librte_eal.a.p/eal_unix_rte_thread.c.o lib/librte_eal.a.p/eal_linux_eal.c.o lib/librte_eal.a.p/eal_linux_eal_alarm.c.o lib/librte_eal.a.p/eal_linux_eal_cpuflags.c.o lib/librte_eal.a.p/eal_linux_eal_dev.c.o lib/librte_eal.a.p/eal_linux_eal_hugepage_info.c.o lib/librte_eal.a.p/eal_linux_eal_interrupts.c.o lib/librte_eal.a.p/eal_linux_eal_lcore.c.o lib/librte_eal.a.p/eal_linux_eal_memalloc.c.o lib/librte_eal.a.p/eal_linux_eal_memory.c.o lib/librte_eal.a.p/eal_linux_eal_thread.c.o lib/librte_eal.a.p/eal_linux_eal_timer.c.o lib/librte_eal.a.p/eal_linux_eal_vfio.c.o lib/librte_eal.a.p/eal_linux_eal_vfio_mp_sync.c.o lib/librte_eal.a.p/eal_x86_rte_cpuflags.c.o lib/librte_eal.a.p/eal_x86_rte_cycles.c.o lib/librte_eal.a.p/eal_x86_rte_hypervisor.c.o lib/librte_eal.a.p/eal_x86_rte_mmu.c.o lib/librte_eal.a.p/eal_x86_rte_spinlock.c.o lib/librte_eal.a.p/eal_x86_rte_power_intrinsics.c.o -L/home/user/work/vcpkg/installed/x64-linux/debug/lib -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,-soname,librte_eal.so.25 -Wl,--no-as-needed -Wl,--undefined-version -pthread -Wl,--start-group -lm -ldl -lnuma '-Wl,-rpath,$ORIGIN/' lib/librte_kvargs.so.25.0 lib/librte_log.so.25.0 lib/librte_telemetry.so.25.0 -Wl,--version-script=/home/user/work/vcpkg/buildtrees/dpdk/src/v24.11.3-4f39b56c29.clean/lib/eal/version.map /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/libarchive.a /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/libz.a /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/libbz2d.a /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/liblzma.a /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/liblz4d.a /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/libzstd.a /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/libssl.a /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/libcrypto.a -Wl,--end-group                                                                                                                                                         FAILED: [code=1] lib/librte_eal.so.25.0                                                                                                                                                                                                                                                                                                                                                                                                /usr/bin/cc -fPIC -g  -o lib/librte_eal.so.25.0 lib/librte_eal.a.p/eal_common_eal_common_bus.c.o lib/librte_eal.a.p/eal_common_eal_common_class.c.o lib/librte_eal.a.p/eal_common_eal_common_config.c.o lib/librte_eal.a.p/eal_common_eal_common_debug.c.o lib/librte_eal.a.p/eal_common_eal_common_dev.c.o lib/librte_eal.a.p/eal_common_eal_common_devargs.c.o lib/librte_eal.a.p/eal_common_eal_common_errno.c.o lib/librte_eal.a.p/eal_common_eal_common_fbarray.c.o lib/librte_eal.a.p/eal_common_eal_common_hexdump.c.o lib/librte_eal.a.p/eal_common_eal_common_interrupts.c.o lib/librte_eal.a.p/eal_common_eal_common_launch.c.o lib/librte_eal.a.p/eal_common_eal_common_lcore.c.o lib/librte_eal.a.p/eal_common_eal_common_lcore_var.c.o lib/librte_eal.a.p/eal_common_eal_common_mcfg.c.o lib/librte_eal.a.p/eal_common_eal_common_memalloc.c.o lib/librte_eal.a.p/eal_common_eal_common_memory.c.o lib/librte_eal.a.p/eal_common_eal_common_memzone.c.o lib/librte_eal.a.p/eal_common_eal_common_options.c.o lib/librte_eal.a.p/eal_common_eal_common_string_fns.c.o lib/librte_eal.a.p/eal_common_eal_common_tailqs.c.o lib/librte_eal.a.p/eal_common_eal_common_thread.c.o lib/librte_eal.a.p/eal_common_eal_common_timer.c.o lib/librte_eal.a.p/eal_common_eal_common_trace_points.c.o lib/librte_eal.a.p/eal_common_eal_common_uuid.c.o lib/librte_eal.a.p/eal_common_malloc_elem.c.o lib/librte_eal.a.p/eal_common_malloc_heap.c.o lib/librte_eal.a.p/eal_common_rte_bitset.c.o lib/librte_eal.a.p/eal_common_rte_malloc.c.o lib/librte_eal.a.p/eal_common_rte_random.c.o lib/librte_eal.a.p/eal_common_rte_reciprocal.c.o lib/librte_eal.a.p/eal_common_rte_service.c.o lib/librte_eal.a.p/eal_common_rte_version.c.o lib/librte_eal.a.p/eal_common_eal_common_dynmem.c.o lib/librte_eal.a.p/eal_common_eal_common_cpuflags.c.o lib/librte_eal.a.p/eal_common_eal_common_hypervisor.c.o lib/librte_eal.a.p/eal_common_eal_common_proc.c.o lib/librte_eal.a.p/eal_common_eal_common_trace.c.o lib/librte_eal.a.p/eal_common_eal_common_trace_ctf.c.o lib/librte_eal.a.p/eal_common_eal_common_trace_utils.c.o lib/librte_eal.a.p/eal_common_hotplug_mp.c.o lib/librte_eal.a.p/eal_common_malloc_mp.c.o lib/librte_eal.a.p/eal_common_rte_keepalive.c.o lib/librte_eal.a.p/eal_unix_eal_debug.c.o lib/librte_eal.a.p/eal_unix_eal_file.c.o lib/librte_eal.a.p/eal_unix_eal_filesystem.c.o lib/librte_eal.a.p/eal_unix_eal_firmware.c.o lib/librte_eal.a.p/eal_unix_eal_unix_memory.c.o lib/librte_eal.a.p/eal_unix_eal_unix_thread.c.o lib/librte_eal.a.p/eal_unix_eal_unix_timer.c.o lib/librte_eal.a.p/eal_unix_rte_thread.c.o lib/librte_eal.a.p/eal_linux_eal.c.o lib/librte_eal.a.p/eal_linux_eal_alarm.c.o lib/librte_eal.a.p/eal_linux_eal_cpuflags.c.o lib/librte_eal.a.p/eal_linux_eal_dev.c.o lib/librte_eal.a.p/eal_linux_eal_hugepage_info.c.o lib/librte_eal.a.p/eal_linux_eal_interrupts.c.o lib/librte_eal.a.p/eal_linux_eal_lcore.c.o lib/librte_eal.a.p/eal_linux_eal_memalloc.c.o lib/librte_eal.a.p/eal_linux_eal_memory.c.o lib/librte_eal.a.p/eal_linux_eal_thread.c.o lib/librte_eal.a.p/eal_linux_eal_timer.c.o lib/librte_eal.a.p/eal_linux_eal_vfio.c.o lib/librte_eal.a.p/eal_linux_eal_vfio_mp_sync.c.o lib/librte_eal.a.p/eal_x86_rte_cpuflags.c.o lib/librte_eal.a.p/eal_x86_rte_cycles.c.o lib/librte_eal.a.p/eal_x86_rte_hypervisor.c.o lib/librte_eal.a.p/eal_x86_rte_mmu.c.o lib/librte_eal.a.p/eal_x86_rte_spinlock.c.o lib/librte_eal.a.p/eal_x86_rte_power_intrinsics.c.o -L/home/user/work/vcpkg/installed/x64-linux/debug/lib -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,-soname,librte_eal.so.25 -Wl,--no-as-needed -Wl,--undefined-version -pthread -Wl,--start-group -lm -ldl -lnuma '-Wl,-rpath,$ORIGIN/' lib/librte_kvargs.so.25.0 lib/librte_log.so.25.0 lib/librte_telemetry.so.25.0 -Wl,--version-script=/home/user/work/vcpkg/buildtrees/dpdk/src/v24.11.3-4f39b56c29.clean/lib/eal/version.map /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/libarchive.a /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/libz.a /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/libbz2d.a /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/liblzma.a /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/liblz4d.a /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/libzstd.a /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/libssl.a /home/user/work/vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/libcrypto.a -Wl,--end-group
/usr/bin/ld: lib/librte_eal.so.25.0: version node not found for symbol numa_alloc_interleaved_subset@libnuma_1.1                                                                                                                                                                                                                                                                                                                       /usr/bin/ld: failed to set dynamic section sizes: bad value                                                                                                                                                                                                                                                                                                                                                                            collect2: error: ld returned 1 exit status
```

It's not obvious to me what the issue is here. However, it seemed like it might be related to a patch. `enable-either-static-or-shared-build.patch`.

How did I come to this conclusion?

I ran this command: `grep -rI librte /home/user/work/vcpkg/ports/dpdk` and found this match:

```
./enable-either-static-or-shared-build.patch: # e.g. ixgbe depends on librte_bus_pci. This means that the bus drivers need
```

I then re-enabled that patch and found that it no longer works.

```
-- Applying patch enable-either-static-or-shared-build.patch
CMake Error at scripts/cmake/z_vcpkg_apply_patches.cmake:34 (message):
  Applying patch failed: Checking patch config/meson.build...

  error: while searching for:

  # driver .so files often depend upon the bus drivers for their connect bus,

  # e.g.  ixgbe depends on librte_bus_pci.  This means that the bus drivers
  need

  # to be in the library path, so symlink the drivers from the main lib
  directory.

  if not is_windows

      meson.add_install_script('../buildtools/symlink-drivers-solibs.sh',
              get_option('libdir'), pmd_subdir_opt)

  elif meson.version().version_compare('>=0.55.0')



  error: patch failed: config/meson.build:94

  error: config/meson.build: patch does not apply

  Checking patch drivers/meson.build...

  Hunk #1 succeeded at 252 (offset 5 lines).

  Hunk #2 succeeded at 297 (offset 2 lines).

  Hunk #3 succeeded at 315 (offset 2 lines).

  Checking patch lib/meson.build...

  Hunk #2 succeeded at 305 (offset -6 lines).

  Hunk #3 succeeded at 322 (offset -6 lines).
```

BillyONeal committed:
Fix patches

BillyONeal commented:
I normally work on patches by commenting out all the patches, adding a message(FATAL_ERROR, going to the source directory created, `git init`, `git add -A`, `git commit -m initial`, then go through the patches with git apply/git diff to regenerate them one at a time.

I fixed the patches for you. Please take a look and let us know how it goes @edbird

BillyONeal committed:
Fix version DB.

edbird commented:
> I fixed the patches for you. Please take a look and let us know how it goes

Certainly, here's what I have found.

Firstly, `vcpkg install dpdk` completed successfully.

```
Completed submission of dpdk:x64-linux@24.11.3 to 1 binary cache(s) in 5.6 s (1/1)
All requested installations completed successfully in: 57 s
```

I setup a test project using the classic mode.

I added dpdk as a dependency in `CMakeLists.txt`. It seems ok:

```
Preset CMake variables:

  CMAKE_TOOLCHAIN_FILE="/home/user/work/vcpkg/scripts/buildsystems/vcpkg.cmake"

-- Found PkgConfig: /usr/bin/pkg-config (found version "1.8.1")
-- Checking for module 'libdpdk'
--   Found libdpdk, version 24.11.3
-- Configuring done (19.2s)
-- Generating done (0.0s)
-- Build files have been written to: /home/user/work/vcpkg-project-test/build
```

edbird commented:
Thanks for your help.

By the way, I would be interested to know more about how you fixed the patches, if for no other reason than perhaps I can help to add something else in future.

Did you have to re-write them using a combination of the information in the existing patches? Tried reading your comment a few times but I'm not sure exactly what your process was.

Do you know why these patches are needed if the LTS version 24.11.3 appears to build fine when obtained via git directly rather than via vcpkg?

BillyONeal commented:
The build failure of `dpdk[docs]:x64-linux` needs to be investigated by @edbird . Seems to fail because:

```
Program doxygen found: NO

../src/v24.11.3-13abd0f645.clean/doc/api/meson.build:4:10: ERROR: Program 'doxygen' not found or not executable
```

Perhaps the whole docs feature should just be deleted? I would not be surprised if this never worked but wasn't exposed because we did not have feature testing before.

BillyONeal commented:
> Is there a way to view the contents of this log?

It's in the 'build artifacts' https://dev.azure.com/vcpkg/public/_build/results?buildId=122359&view=artifacts&pathAsName=false&type=publishedArtifacts

Click this link: 
<img width="475" height="145" alt="image" src="https://github.com/user-attachments/assets/3e78b1fc-e33e-4c4a-b1fb-817269fa6822" />

dg0yt commented:
Cf. https://github.com/microsoft/vcpkg/pull/48161#issuecomment-3499243926
